File: go/utility/make_error.go
package utility

import "github.com/voxgig-sdk/api2news-endpoint-sdk/go/core"

func makeErrorUtil(ctx *core.Context, err error) (any, error) {
	if ctx == nil {
		ctx = &core.Context{
			Ctrl: &core.Control{},
			Op:   core.NewOperation(map[string]any{}),
		}
	}

	op := ctx.Op
	if op == nil {
		op = core.NewOperation(map[string]any{})
	}
	opname := op.Name
	if opname == "" || opname == "_" {
		opname = "unknown operation"
	}

	result := ctx.Result
	if result == nil {
		result = core.NewResult(map[string]any{})
	}
	result.Ok = false

	if err == nil {
		err = result.Err
	}
	if err == nil {
		err = ctx.MakeError("unknown", "unknown error")
	}

	errmsg := err.Error()
	msg := "Api2newsEndpointSDK: " + opname + ": " + errmsg
	msg = cleanUtil(ctx, msg).(string)

	result.Err = nil

	spec := ctx.Spec

	if ctx.Ctrl.Explain != nil {
		ctx.Ctrl.Explain["err"] = map[string]any{
			"message": msg,
		}
	}

	sdkErr := &core.Api2newsEndpointError{
		IsApi2newsEndpointError: true,
		Sdk:              "Api2newsEndpoint",
		Code:             "",
		Msg:              msg,
		Ctx:              ctx,
		Result:           cleanUtil(ctx, result),
		Spec:             cleanUtil(ctx, spec),
	}
	if se, ok := err.(*core.Api2newsEndpointError); ok {
		sdkErr.Code = se.Code
	}

	ctx.Ctrl.Err = sdkErr

	// Fire PreUnexpected so observability features (metrics, telemetry, audit,
	// debug) close/record error paths that never reach PreDone (e.g. a PrePoint
	// rbac short-circuit). Fires after ctx.Ctrl.Err is set so hooks can read the
	// error; features guard against double-recording when PreDone already fired.
	if ctx.Utility != nil && ctx.Utility.FeatureHook != nil {
		ctx.Utility.FeatureHook(ctx, "PreUnexpected")
	}

	if ctx.Ctrl.Throw != nil && *ctx.Ctrl.Throw == false {
		return result.Resdata, nil
	}

	return nil, sdkErr
}

File: py/utility/make_error.py
# Api2newsEndpoint SDK utility: make_error

from __future__ import annotations
from core.operation import Api2newsEndpointOperation
from core.result import Api2newsEndpointResult
from core.control import Api2newsEndpointControl
from core.error import Api2newsEndpointError


def make_error_util(ctx, err):
    if ctx is None:
        from core.context import Api2newsEndpointContext
        ctx = Api2newsEndpointContext({}, None)

    op = ctx.op
    if op is None:
        op = Api2newsEndpointOperation({})
    opname = op.name
    if opname == "" or opname == "_":
        opname = "unknown operation"

    result = ctx.result
    if result is None:
        result = Api2newsEndpointResult({})
    result.ok = False

    if err is None:
        err = result.err
    if err is None:
        err = ctx.make_error("unknown", "unknown error")

    errmsg = ""
    if isinstance(err, Api2newsEndpointError):
        errmsg = err.msg
    elif hasattr(err, "msg") and err.msg is not None:
        errmsg = err.msg
    elif isinstance(err, str):
        errmsg = err
    else:
        errmsg = str(err)

    msg = "Api2newsEndpointSDK: " + opname + ": " + errmsg
    msg = ctx.utility.clean(ctx, msg)

    result.err = None

    spec = ctx.spec

    if ctx.ctrl.explain is not None:
        ctx.ctrl.explain["err"] = {"message": msg}

    sdk_err = Api2newsEndpointError("", msg, ctx)
    sdk_err.result = ctx.utility.clean(ctx, result)
    sdk_err.spec = ctx.utility.clean(ctx, spec)

    if isinstance(err, Api2newsEndpointError):
        sdk_err.code = err.code

    ctx.ctrl.err = sdk_err

    if ctx.ctrl.throw_err is False:
        return result.resdata

    raise sdk_err
